services/feed-manager-service/feed-manager-controller/src/main/java/com/thinkbiganalytics/feedmgr/rest/controller/FeedRestController.java

```diff
@@ -858,22 +860,95 @@ public DatasourceDefinitions updateDatasourceDefinitions(DatasourceDefinitions d
+    public Response uploadFile(@PathParam("feedId") String feedId, FormDataMultiPart multiPart) {
+        List<NifiProperty> properties = getNifiProperties(feedId);
+        FileUploadContext context = getFileUploadContext(properties);
 
-        FeedMetadata feed = getMetadataService().getFeedById(feedId, false);
-        // Derive path and file
-        feed = registeredTemplateService.mergeTemplatePropertiesWithFeed(feed);
-        propertyExpressionResolver.resolvePropertyExpressions(feed);
-        List<NifiProperty> properties = feed.getProperties();
+        if (!context.isValid()) {
+            throw new InternalServerErrorException("Unable to upload file with empty dropzone or file");
+        }
+
+        List<BodyPart> bodyParts = multiPart.getBodyParts();
+        List<String> uploadedFiles = new ArrayList<>();
+        for (BodyPart bodyPart : bodyParts) {
+            BodyPartEntity entity = (BodyPartEntity)bodyPart.getEntity();
+            String fileName = bodyPart.getContentDisposition().getFileName();
+            try {
+                saveFile(entity.getInputStream(), context);
+                uploadedFiles.add(fileName);
+            } catch (AccessDeniedException e) {
+                String errTemplate = getErrorTemplate(uploadedFiles, "Permission denied attempting to write file [%s] to [%s]. Check with system administrator to ensure this application has write permissions to folder");
+                String err = String.format(errTemplate, fileName, context.getDropzone());
+                log.error(err);
+                throw new InternalServerErrorException(err);
+
+            } catch (Exception e) {
+                String errTemplate = getErrorTemplate(uploadedFiles, "Unexpected exception writing file [%s] to [%s].");
+                String err = String.format(errTemplate, fileName, context.getDropzone());
+                log.error(err);
+                throw new InternalServerErrorException(err, e);
+            }
+        }
+
+        return Response.ok("").build();
+    }
+
+    private String getErrorTemplate(List<String> uploadedFiles, String msg) {
+        return uploadedFiles.size() > 0 ? "Successfully uploaded files: " + String.join(", ", uploadedFiles) + ".\n" + msg : msg;
+    }
+
+    private void saveFile(InputStream stream, FileUploadContext context) throws IOException {
+        File tempTarget = File.createTempFile("kylo-upload", "");
+
+        java.nio.file.Path dropZoneTarget = generateDropZonePath(context);
+        if (dropZoneTarget == null) {
+            throw new IOException("Unable to upload file");
+        }
+        File dropZoneFile = dropZoneTarget.toFile();
+
+        Files.copy(stream, tempTarget.toPath(), StandardCopyOption.REPLACE_EXISTING);
+        Files.move(tempTarget.toPath(), dropZoneTarget);
+
+        // Set read, write
+        dropZoneFile.setReadable(true);
+        dropZoneFile.setWritable(true);
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Do something with the "boolean" value returned by "setWritable". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS899)